### PR TITLE
fix: make refresh_unity's compile wait observable across the domain reload

### LINKED ISSUE #814: Agents sleep after script changes
# Problem

When an AI agent (e.g. in Cursor) edits a script and then needs to play the game or run other Unity operations, it must wait for Unity to finish recompiling. Today there is no dedicated way to “wait until compilation is done,” so agents tend to:

1. **Use fixed sleeps** (e.g. `sleep 5`, `sleep 3`) after script changes, which is slow and brittle (sometimes compile is done in 1s, sometimes longer).
2. **Poll** the `editor_state` resource for `compilation.is_compiling` in a loop, which works but is verbose and easy to get wrong.

In a real “play → read console → fix error → re-verify” workflow, this led to roughly **~12 seconds of sleep** in a single run (play wait, recompile wait, play-again wait), making the loop feel slow even when the actual fix was simple.

## Proposed solution

Add a **`wait_for_compilation`** capability so the agent can explicitly wait for compilation (and domain reload) to finish instead of guessing with sleeps or implementing custom polling.

Concretely:

- **`manage_editor(action="wait_for_compilation")`**  
  - Server-side only (no new Unity C# API required).  
  - Uses existing `editor_state` / compilation state and the existing “wait for editor ready” logic (e.g. `wait_for_editor_ready` / `wait_for_no_compile`) to block until compilation is done.  
  - Optional `timeout` (e.g. default 30s, max 120s).  
  - Returns as soon as compilation is already done, or after it finishes within the timeout.

This keeps the workflow clear for the agent:

1. `manage_script(action="update", ...)` — apply script change  
2. `manage_editor(action="wait_for_compilation")` — wait until Unity is ready  
3. `manage_editor(action="play")` — enter play mode  
4. `read_console(...)` — check for errors  

No `play_and_check` or other new high-level action is required for this step; the main win is removing the need for long, fixed sleeps and ad-hoc polling.

## References

- Here is my chat log

Image
Image

### PR DESCRIPTION
**Pull request by KamilDev**
## Description

`refresh_unity(compile="request", wait_for_ready=true)` returns before the compile it requested has started, and that compile's own edges are erased by the domain reload that ends it. Agents fall back to fixed sleeps as a result — which is #814.

Two independent causes, both measured on Unity 6000.3.14f1:

**1. The readiness wait races the start of the compile.** `CompilationPipeline.RequestScriptCompilation()` only queues; the pipeline starts on a later editor tick. `resulting_state` is sampled immediately after, so it reports `idle` for a compile that is about to run — and the server-side `wait_for_editor_ready` poll, which begins the moment the tool returns, sees a ready editor and returns at once. `wait_for_ready` therefore does nothing for exactly the call it exists for.

Instrumented with a `SessionState`-backed probe on the real `compilationStarted` event:

```
refresh_unity  ->  {"resulting_state": "idle"}     # returned immediately
compilationStarted fired 3706 ms later
```

**2. The compile's edges do not survive the reload that ends it.** `last_compile_started_unix_ms` / `last_compile_finished_unix_ms` were derived by edge-detecting `GetActualIsCompiling()` on the throttled update tick, into statics. A *successful* compile ends in a domain reload that wipes them, so the falling edge of the very compile a client is waiting on is unobservable — both fields read `null` afterwards, leaving "finished" and "never started" indistinguishable. Straight after a compile that demonstrably ran and reloaded the domain:

```json
"compilation": {
  "last_compile_started_unix_ms": null,
  "last_compile_finished_unix_ms": null,
  "last_domain_reload_after_unix_ms": 1787707049894
}
```

The sampling also quantised both values to the 1s update throttle, and missed any compile shorter than one tick.

## Type of Change
- [x] Bug fix (non-breaking change that fixes an issue)

## Changes Made

**`EditorStateCache.cs`** — record the compile edges from `CompilationPipeline.compilationStarted` / `compilationFinished` into `SessionState` instead of sampling them into statics. `compilationFinished` fires *before* the reload, so the write lands while the domain is alive and is read back by the next one. `SessionState` survives reloads and dies with the editor session, which is the lifetime these values describe. The events were already subscribed for `GetActualIsCompiling` — only the storage changes. Adds an internal monotonic `CompileCount` alongside them.

**`RefreshUnity.cs`** — when `wait_for_ready` is true, wait for the start edge before reporting state, so `resulting_state`, and every readiness decision downstream of it, is truthful. `CompileCount` backs the wait so it also catches a compile that begins and ends inside `AssetDatabase.Refresh`, before the wait is armed. Bounded by a 10s grace and *resolved*, never faulted, when the pipeline never starts. The outcome is surfaced as `compile_started` (`true` / `false`, or `null` when nothing was waited for). `wait_for_ready=false` stays non-blocking: no wait, immediate return, poll hint as before.

Unlike `WaitForUnityReadyAsync`, this wait cannot span a domain reload — it returns the moment compilation *starts*, long before assemblies swap. The Unity 6+ opt-out guarding the readiness wait (`waitForReady && !compileRequested`) is left exactly as it is and does not apply here.

The start is observed from `CompilationPipeline.compilationStarted` rather than an update poll, and the completion source deliberately omits `RunContinuationsAsynchronously`. Every `await` between the wait and the socket captures Unity's synchronization context, and a continuation posted to it runs one editor frame later; there are three such hops (this handler, the `CommandRegistry` async wrapper, its `AwaitHandler`). Completed on the main thread with inlining allowed, all three run inside the event handler and the response reaches the thread-pool send within the compile, which is what keeps a ~110 ms cached compile from reloading over the response.

No schema change and no server-side change: `CompileCount` stays internal to the package, and the existing `wait_for_editor_ready` loop now works because the state it polls is finally truthful. `compile_started` is one additional field on the response payload.

On when the grace actually fires — review raised that `RequestScriptCompilation` "recompiles those scripts which require it" and might therefore not start at all when nothing changed, making the grace the normal path. Measured, it does not work that way: with nothing changed, 6000.3.14f1 still raises `compilationStarted`/`compilationFinished` (a ~110 ms cached run) and reloads the domain, and the counter moves on every call. In `UnityCsReference`, `EditorCompilation.RequestScriptCompilation` records a pending request that native drains into `CompileScriptsWithSettings` unconditionally, identically on the `2021.3` branch and `master`; the per-assembly skipping is inside that run. The grace covers a run that never begins — a compilation setup error, or play mode with "Recompile After Finished Playing" — and both now report `compile_started=false` instead of just being slow.

## Compatibility / Package Source
- Unity version(s) tested: 6000.3.14f1
- Package source used: `file:` (local checkout, live-linked into a 6000.3.14f1 project)
- Resolved commit hash: n/a (`file:` source)

## Testing/Screenshots/Recordings
- [x] Python tests (`cd Server && uv run pytest tests/ -v`) — 1374 passed, 3 skipped. Unchanged by this PR; run to confirm no regression.
- [x] Unity EditMode tests — `RefreshUnityTests` (3) and `EditorStateCacheSessionValuesTests` (5) ran as NUnit fixtures on 6000.3.14f1 through the bridge's `run_tests`, 8/8 passed. They cover the synchronous exits, the grace expiry, and the `SessionState` helpers; they do not drive a domain reload.
- [ ] Unity PlayMode tests
- [x] Package import/compile check — `tools/compile-check.sh` green for win/osx/linux against 6000.3.14f1 and 2022.3.27f1 Hub installs.

Verified against a live Editor, reproducing the failure first and then the fix on the same call:

| | before | after |
|---|---|---|
| `resulting_state` | `idle` | `compiling` |
| `last_compile_started_unix_ms` | `null` | `1787707352845` |
| `last_compile_finished_unix_ms` | `null` | `1787707354510` (1665 ms compile) |
| still readable after the reload | — | yes, 5214 ms past `compilationFinished` |

`compile_started`, live on 6000.3.14f1, one call per row:

| call | returned after | `compile_started` | `resulting_state` |
|---|---|---|---|
| `compile="request"`, `wait_for_ready=false`, nothing changed | 50 ms | `null` | `idle` |
| `compile="request"`, `wait_for_ready=true`, source changed (1007 ms compile) | 1326 ms | `true` | `compiling` |
| `compile="request"`, `wait_for_ready=true`, nothing changed (~110 ms cached compile), raw bridge, 3 runs | ~0.8 s | `true` | `compiling` |
| `compile="request"`, `wait_for_ready=true`, in play mode with "Recompile After Finished Playing" | 10.01 s | `false` | `idle` (counter unchanged; the compile ran on play exit) |

The third row is the one the branch initially broke. Waiting for the start edge moves the response into the window between `compilationStarted` and the reload, and with the wait observed from an update poll plus three context-posted continuations between it and the socket, a ~110 ms cached compile beat the response every time (4/4 lost; the MCP tool masked it as `recovered_from_disconnect`, the CLI printed a disconnect error). Resolving from the `compilationStarted` event with continuations inlined closed it — see the last commit.

**Not covered by a fixture:** the reload-spanning behavior itself. It is a timing interaction with the real compilation pipeline and domain reload, so it was exercised against a live Editor rather than asserted in a test — I did not want to imply fixture coverage that does not exist.

## Documentation Updates
- [ ] I have added/removed/modified tools or resources

No tool or resource surface changes — same tool, same parameters. The response gains one field, `compile_started`; the reference docs describe the payload as action-dependent and list no fields, so nothing to regenerate. `last_compile_started_unix_ms` / `last_compile_finished_unix_ms` are already documented; they now actually hold values.

## Related Issues

Fixes #814.

Relates to #978, which adds `manage_editor(action="wait_for_compilation")` for the same issue. That PR delegates to `wait_for_editor_ready` and touches no C#, so it inherits both causes above — its own first test (`test_wait_for_compilation_returns_immediately_when_ready`) passes identically whether compilation finished or never began. This PR fixes the state that call depends on, so the two are complementary rather than competing: with this landed, #978's wrapper would do what its name says.

Also relevant to #1276 and #549, which are earlier instances of the same underlying pattern — trusting a sampled flag over the compilation events.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved compilation state tracking across Unity domain reloads.
  - Compilation events now update snapshots reliably and preserve timestamps.
  - Added a short grace period to prevent premature readiness reports when compilation begins.
  - Improved detection of active and recently completed compilation.
  - Ensured temporary monitoring is cleaned up safely, including after errors.
  - Improved refresh readiness reporting for more reliable editor synchronization.

- **Tests**
  - Added coverage for compilation tracking, refresh completion, and readiness reporting scenarios.

### CODE REVIEW

File: MCPForUnity/Editor/Tools/RefreshUnity.cs [hunk 1]
```diff
@@ -124,6 +149,61 @@ await WaitForUnityReadyAsync(
...
+        /// <item><see cref="EditorStateCache.CompileCount"/> moved past
+        /// <paramref name="compileCountBefore"/> — a short compile can begin and end
+        /// inside AssetDatabase.Refresh, before this is even armed, and the counter is
+        /// the only thing that still sees it;</item>
+        /// <item>the grace elapsed with neither.</item>
+        /// </list>
+        /// </summary>
+        private static Task WaitForCompilationToStartAsync(int compileCountBefore, TimeSpan grace)
+        {
+            var tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
```
[the remark this answers is not in the document]
Comment (KamilDev) [reply MCPForUnity/Editor/Tools/RefreshUnity.cs]: Good catch on the branch that matters — fixed in e039964a.

You're right about the counter branch specifically. `EditorStateCache.CompileCount != compileCountBefore` exists for a compile that begins *and ends* inside `AssetDatabase.Refresh`, so it can resolve with the reload already imminent, and resolving it from `Tick` hands the rest of `HandleCommand` to the synchronization context as a queued continuation. My PR description's claim that this wait "cannot span a domain reload" held for the `isCompiling` branch but not that one.

Fix is a synchronous test of both exit conditions on entry, returning a completed task so the await resumes inline and nothing is left queued:

```csharp
if (EditorStateCache.CompileCount != compileCountBefore
    || EditorStateCache.GetActualIsCompiling())
{
    return Task.CompletedTask;
}
```

The polling path is now reached only when no compile has started yet — where the reload is at minimum a compile away, not one tick.

Two clarifications on the rest of the finding, for the record rather than as disagreement:

- The `isCompiling` branch resolves while compilation is *running*, so the reload is a full compile away. Measured on 6000.3.14f1: 1668 ms compile, reload 5.4 s after `compilationFinished`. That branch fired in every live run I made.
- A reload eating a response is not a new failure mode this introduces, and it doesn't surface as a bare timeout: the server already treats connection loss after send as expected during a compile and waits for readiness (`is_connection_lost_after_send` → `recovered_from_disconnect` in `Server/src/services/tools/refresh_unity.py`). I saw that path fire during verification — it returned `"Refresh recovered after Unity disconnect/retry; editor is ready."` rather than an error.

Re-verified after the change: `resulting_state: "compiling"`, `last_compile_started/finished_unix_ms` populated (1668 ms compile) and still readable 5435 ms later, past the reload. `tools/compile-check.sh` green for win/osx/linux.

**Comment by Scriptwonder** [comment]
Thanks for the effort here, a few notes (from AI audit):

The comment on `CompileStartGraceSeconds` says `RequestScriptCompilation` always runs a pass, but the 2021.3 and 6000.0 docs both say it "recompiles those scripts which require it", which is also why `CleanBuildCache` exists.
So when nothing needs recompiling, `CompileCount` never moves and the wait burns the full
10s. That makes the grace the normal path rather than a backstop, and it lands on the
common case: a no-op `refresh_unity(compile="request")` now costs 10s. Because the new
block sits above `shouldWaitForReady`, it costs that even with `wait_for_ready=false`,
which we document as non-blocking. Play mode with Recompile After Finished Playing
behaves the same way.

Most of the fix is already there. `WaitForCompilationToStartAsync` sets true/false for
exactly this case, but it's typed `Task` and the caller drops the result. Return
`Task<bool>` and surface it as `compile_started` in the payload, so "nothing needed
compiling" is visible instead of just slow. Gating the wait on `waitForReady` as well
would keep the non-blocking path honest.

Could you add a couple of tests while you're in there? `compile="none"` returning an
already-completed task, and `GetSessionUnixMs` round-tripping and returning null on a
malformed value, cover the new code without driving a reload.

**Comment by KamilDev** [comment]
Thanks — all four are in 229b6117 (rebased onto current `beta`, only the two files touched).

- `WaitForCompilationToStartAsync` returns `Task<bool>`, surfaced as `compile_started`. `null` when nothing was waited for (no compile requested, or `wait_for_ready=false`), so "not observed" never reads as "did not start".
- The wait is gated on `wait_for_ready`. The non-blocking path returns in ~50 ms with `compile_started: null` and the poll hint.
- Tests: `compile="none"` with `wait_for_ready=false` hands back an already-completed task; a moved counter resolves the wait synchronously with `true`; `GetSessionUnixMs` round-trips a value beyond `int.MaxValue` and returns `null` for unset and malformed input. They ran as NUnit fixtures on 6000.3.14f1 (7/7). The `compile="none"` test tolerates the `tests_running` short-circuit, since that is the branch it takes under the bridge's `run_tests`.
- The `CompileStartGraceSeconds` doc was wrong in the other direction, so I rewrote it rather than softening it.

On the premise, since the audit read the docs literally: I measured it before changing anything. Four `compile="request"` calls with nothing changed on 6000.3.14f1 each moved `CompileCount`, fired `compilationStarted`/`compilationFinished` (~110 ms, Bee cache), and reloaded the domain. That matches the source: `EditorCompilation.RequestScriptCompilation` sets `m_ScriptCompilationRequest`, and native drains it into `CompileScriptsWithSettings` unconditionally — identical on the `2021.3` branch and `master`. "Recompiles those scripts which require it" is per-assembly skipping inside that run (`assemblyCompilationNotRequired` is the event for it), not a run that is skipped. So the grace was never the normal path for a no-op request; the start edge arrives within a tick.

The case where it *is* the normal path is the play-mode one you named. With "Recompile After Finished Playing" and play mode active, the request is deferred until play exits: measured 10.01 s and `compile_started: false`, with the counter unchanged until play stopped. That call now says so instead of just being slow, and `wait_for_ready=false` skips it entirely.

One thing worth knowing that this branch does not change: with a no-op request the reload lands before the response leaves the socket, so the raw bridge sees a disconnect for that call. The MCP tool already maps that to `recovered_from_disconnect` and polls readiness, so agents see success — the payload is what gets lost on that path. Noted in the PR body.

File: MCPForUnity/Editor/Tools/RefreshUnity.cs [hunk 2]
```diff
@@ -77,6 +90,34 @@ public static async Task<object> HandleCommand(JObject @params)
...
+            //
+            // compile_started is null when nothing was waited for (no compile
+            // requested, or wait_for_ready=false), so "not observed" never reads as
+            // "did not start".
+            bool? compileStarted = null;
+            if (compileRequested && waitForReady)
+            {
+                compileStarted = await WaitForCompilationToStartAsync(
+                    compileCountBefore,
+                    TimeSpan.FromSeconds(CompileStartGraceSeconds)).ConfigureAwait(true);
```
[the remark this answers is not in the document]
Comment (KamilDev) [reply MCPForUnity/Editor/Tools/RefreshUnity.cs:118]: The premise does not hold on 6000.3.14f1, and I measured it on that exact version before this round rather than reading the docs. Four `compile="request"` calls with nothing changed each moved `CompileCount`, fired `compilationStarted`/`compilationFinished` (~110 ms, Bee cache) and reloaded the domain; three more after 19966ba5 did the same and returned `compile_started: true` in ~0.8 s. In `UnityCsReference`, `EditorCompilation.RequestScriptCompilation` sets `m_ScriptCompilationRequest` and native drains it into `CompileScriptsWithSettings` unconditionally, on the `2021.3` branch and `master` alike — "recompiles those scripts which require it" is per-assembly skipping inside that run (`assemblyCompilationNotRequired` is its event), not a run that is skipped. So there is no no-work case to detect: the start edge arrives within a tick.

`RequestScriptCompilationOptions.CleanBuildCache` would go the other way — it deletes `Library/Bee` artifacts and forces a full rebuild on every refresh, turning a ~110 ms cached run into a cold one.

The grace only fires when the run never begins — a compilation setup error, or play mode with "Recompile After Finished Playing" deferring the request until exit (measured 10.01 s, `compile_started: false`). That is the honest outcome for those calls, and `wait_for_ready=false` skips the wait entirely.

**Comment by KamilDev** [comment]
Correction to the last paragraph of my previous comment, plus a fix: the lost response was *not* pre-existing. Before this branch the response left before the compile even started; waiting for the start edge moved it into the window between `compilationStarted` and the reload, and for a cached no-change compile (~110 ms) that window was too short. The wait was observed from an update poll, and each of the three `await ... ConfigureAwait(true)` hops between it and the socket ran one editor frame later, so the reload won 4/4 times. The MCP tool masked it as `recovered_from_disconnect`; `unity-mcp editor refresh --compile` printed a disconnect error.

19966ba5 resolves the wait from `CompilationPipeline.compilationStarted` itself, with the completion source completing inline so those three continuations run inside the event handler and only the thread-pool send remains. Re-measured on the same raw-bridge call: 3/3 no-op requests returned `compile_started: true`, `resulting_state: "compiling"` in ~0.8 s. Added a `[UnityTest]` for the grace-expiry path (zero grace resolves `false` on the first tick); the fixtures ran 8/8. PR body updated to match.
